src/Roslyn.Diagnostics.Analyzers/UnitTests/DoNotCopyValueTests.cs

```diff
@@ -0,0 +1,697 @@
+    {{
+        {csharpFieldReference} = GCHandle.Alloc(new object());
+    }}
+}}
+");
+
+            if (visualBasicFieldReference is object)
+            {
+                await VerifyVB.VerifyAnalyzerAsync($@"
+Imports System.Runtime.InteropServices
+
+Class C
+    Dim field As GCHandle
+
+    Sub M()
+        {visualBasicFieldReference} = GCHandle.Alloc(New Object())
+    End Sub
+End Class");
+            }
+        }
+
+        [Theory]
+        [InlineData("field", "field")]
+        [InlineData("(field)", null)]
+        [InlineData("this.field", "Me.field")]
+        [InlineData("(this).field", null)]
+        [InlineData("((C)this).field", "DirectCast(Me, C).field")]
+        public async Task TestAcquireIntoArrayFieldFromReturnByValue(string csharpFieldReference, string? visualBasicFieldReference)
+        {
+            await VerifyCS.VerifyAnalyzerAsync($@"
+using System.Runtime.InteropServices;
+
+class C
+{{
+    GCHandle[] field;
+
+    void M()
+    {{
+        {csharpFieldReference}[0] = GCHandle.Alloc(new object());
+    }}
+}}
+");
+
+            if (visualBasicFieldReference is object)
+            {
+                await VerifyVB.VerifyAnalyzerAsync($@"
+Imports System.Runtime.InteropServices
+
+Class C
+    Dim field As GCHandle()
+
+    Sub M()
+        {visualBasicFieldReference}(0) = GCHandle.Alloc(New Object())
+    End Sub
+End Class");
+            }
+        }
+
+        [Fact]
+        public async Task TestDoNotAcquireFromReturnByReference()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System.Runtime.InteropServices;
+
+class C
+{
+    void M()
+    {
+        var local = GetRef();
+    }
+
+    ref GCHandle GetRef() => throw null;
+}
+",
+                // /0/Test0.cs(8,21): warning RS0042: Unsupported use of non-copyable type 'System.Runtime.InteropServices.GCHandle' in 'Invocation' operation
+                VerifyCS.Diagnostic(DoNotCopyValue.UnsupportedUseRule).WithSpan(8, 21, 8, 29).WithArguments("System.Runtime.InteropServices.GCHandle", "Invocation"));
+        }
+
+        [Fact]
+        public async Task TestPassToInstancePropertyGetter()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System.Runtime.InteropServices;
+
+class C
+{
+    GCHandle field;
+    readonly GCHandle readonlyfield;
+
+    void M()
+    {
+        var local = new GCHandle();
+        ref var reflocal = ref local;
+        ref readonly var refreadonlylocal = ref local;
+
+        // Call each proprety twice to ensure the analyzer didn't silently treat one like a move
+        _ = field.Target;
+        _ = field.Target;
+        _ = local.Target;
+        _ = local.Target;
+        _ = reflocal.Target;
+        _ = reflocal.Target;
+
+        _ = readonlyfield.Target;
+        _ = refreadonlylocal.Target;
+    }
+}
+",
+                // /0/Test0.cs(23,13): warning RS0042: Unsupported use of non-copyable type 'System.Runtime.InteropServices.GCHandle' in 'FieldReference' operation
```

Comment:
Can we use markup based syntax here to easily map the flagged code?

Reply:
Yes, the new markup features finally allow this to be cleaned up.

Reply:
Done!